soc: nordic: Add driver for Nordic MVDMA peripheral

nordic-krch opened the pull request:
Driver does not implement Zephyr DMA API. The reason is that it will mainly be used in the vendor specific code and translation to the Zephyr API significantly decreases the performance to a point where using DMA is no longer beneficial.

Driver offers free running transfer setup and a transfer with a callback. If free running mode is used (without a callback) then `nrf_mvdma_xfer_check` can be used to poll for transfer completion. It is required to poll until function returns success (transfer has finished). That is because MVDMA blocks entering S2RAM PM mode and unblocking is done only when transfer completes.

Test contains timings measurements that can be used to decide whether use of MVDMA is beneficial.

DNM as it requires #96909.

nordic-krch committed:
soc: nordic: common: Add driver for Nordic MVDMA peripheral

MVDMA driver does not implement Zephyr DMA API. It implements its own
due to performance reasons and it is intended to be used mainly in the
Nordic specific code.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: boards: nrf: Add test for MVDMA driver

Add test suite for MVDMA driver.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
dts: vendor: nordic: Add mvdma nodes

Add MVDMA binding and nodes to nrf54h20.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
soc: nordic: nrf54h: Add MVDMA restore to S2RAM

MVDMA is present in the application domain so it's configuration
need to be restored when waking up from suspend to RAM.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

teburd reviewed (dismissed):
I'm a bit confused by this PR to be honest, it is adding the driver to the DMA drivers, but it is not in fact a DMA driver from Zephyr's perspective. I'm not sure exactly where this should go, but I do not believe it should go into drivers/dma.

To be clear, I'm not at all opposed to a vendor creating their own APIs and drivers around DMA, as noted already in the documentation the DMA API is a bit of a union of DMA peripheral capabilities. If DMA-like actions are wanted in a cross platform way we already provide this in peripheral APIs in part. But to put this driver in drivers/dma should imply it implements the dma.h interface which it does not.

teburd commented:
Maybe to ask more broadly, what about the current DMA interface isn’t workable for you? It’s not meant to provide uniform behavior but a uniform API, DMAs are already quite vendor specific in how they behave

teburd commented:
> > what about the current DMA interface isn’t workable for you?
> 
> @teburd It's not that anything is not working in the API. This low level driver can be used for the driver that will implement Zephyr DMA API. It is just a matter of performance and the fact that it is much more likely that the driver will be used by vendor specific code than by the user application. MVDMA works with job descriptors in RAM so for each `struct dma_block_config` driver would need to allocate a RAM block and translate that request to the MVDMA specific descriptions. In Nordic SoC each peripheral has it's own DMA (so called EasyDMA) which is handled in a peripheral driver so a typical use case for the DMA (memory-peripheral transfers) is handled there already. MVDMA is a helper that can offload CPU in certain cases (e.g. copying data between slow RAM and fast RAM) and the longer is the setup the less cases it will cover.

So to maybe hone in a bit here, it is the translation between the generic dma descriptors to the hardware descriptors that is a concern? I think this is a valid concern.

As I said I have no problem myself with this being in some vendor specific driver location, but it does not belong in drivers/dma to me where I am then labeled as the maintainer.

nashif reviewed (changes requested):
We should not be adding drviers under soc/ that are covered by an existing Zephyr API. This requires further discussion and clarification on why this is needed and why something like this can't be implemented under drivers/ using zephyr APIs.

nordic-krch committed:
pm: policy: Add option to lock all power states

Add function for getting and putting a lock for all power states.
It is much faster version of requesting 0 us latency with actual
intention to disable all power states.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented on include/zephyr/pm/policy.h:
```diff
@@ -125,6 +126,18 @@ void pm_policy_state_lock_get(enum pm_state state, uint8_t substate_id);
  */
 void pm_policy_state_lock_put(enum pm_state state, uint8_t substate_id);
 
+/**
```
it is there: #96909

nordic-krch commented:
> Is this driver in any way accessible to an application or usable by an application?

Currently we don't see a use case. Application operates on "fast" RAM. For memory-to-memory operations time needed to setup a MVDMA will be longer than memory operation itself. nRF54H20 SoC has "slow" RAM that is used by peripherals and main use case for MVDMA is to copy data between "fast/user" memory and "slow" which happens in drivers.

carlescufi commented:
Architecture WG meeting:

- On Nordic devices, there's DMA for each peripheral that is independent of the DMA being exposed here
- The MVDMA engine instead can copy from memory to memory and from register space to memory
- The peripherals themselves can only access slow RAM, but usually buffers are populated in fast RAM, so copying is needed
- MVDMA however can access both fast and slow RAM, and so the purpose of this driver is to be able to copy from and to slow/fast RAM
- Using MVDMA only makes sense if setting up the DMA transfer is faster than actually copying memory, and using the standard DMA API would effectively negate the advantages
  - In particular, preparing the descriptors is a time-consuming task
- Why can't then just be in the HAL?
  - It's using Zephyr APIs, particularly power management and cache APIs
